generate-spec

```diff
@@ -680,18 +714,89 @@ if (count($schemas) == 0 && count($routes) == 0) {
 	Logger::error("app", "No spec generated");
 }
 
-if (count($openapi["paths"]) == 0) {
-	$openapi["paths"] = new stdClass();
-}
-
 ksort($schemas);
-$openapi["components"]["schemas"] = count($schemas) == 0 ? new stdClass() : $schemas;
 
 if ($useTags) {
 	$openapi["tags"] = $tags;
 }
 
-file_put_contents($out, json_encode($openapi, Helpers::jsonFlags()));
+foreach ($scopePaths as $scope => $paths) {
+	$openapiScope = $openapi;
+
+	if (count($paths) == 0) {
+		$paths = new stdClass();
+	}
+
+	$scopeSuffix = $scope === 'default' ? '' : '-' . $scope;
+	$openapiScope['info']['title'] .= $scopeSuffix;
+	$openapiScope['paths'] = $paths;
+
+	$usedSchemas = [];
+	foreach ($paths as $url => $urlRoutes) {
+		foreach ($urlRoutes as $httpMethod => $routeData) {
+			foreach ($routeData['responses'] as $statusCode => $responseData) {
+				if (isset($responseData['content']['application/json'])) {
+					$usedSchemas = array_merge($usedSchemas, Helpers::collectUsedRefs($responseData['content']['application/json']['schema']));
+				} elseif (isset($responseData['content']['*/*']['schema']['type'], $responseData['content']['*/*']['schema']['format'])
+					&& $responseData['content']['*/*']['schema']['type'] === 'string'
+					&& $responseData['content']['*/*']['schema']['format'] === 'binary') {
+					Logger::info("app", "Binary response from '$httpMethod $url' - Skipping schema reading");
+				} else {
+					Logger::warning("app", "Could not read used schemas for response to '$httpMethod $url' with status code $statusCode");
+				}
+			}
+		}
+	}
+
+	$scopedSchemas = [];
+	while ($usedSchema = array_shift($usedSchemas)) {
+		if (!str_starts_with($usedSchema, '#/components/schemas/')) {
+			continue;
+		}
+
+		$schemaName = substr($usedSchema, strlen('#/components/schemas/'));
+
+		if (!isset($schemas[$schemaName])) {
+			Logger::error("app", "Schema $schemaName used by scope $scope is not defined");
+		}
+
+		// Queue potential sub-refs for exporting as well
+		if (isset($schemas[$schemaName]['allOf'])) {
```

Comment:
Only checking allOf is not enough. All schemas need to be traversed and checked for a property `$ref` (I think you already realized that with the array items from the last commit)